File: examples/agent-file/emit.mjs
#!/usr/bin/env node
// Quickstart: emit one valid agent spool file (schema v1).
//
//   node examples/agent-file/emit.mjs /path/to/spool --id demo --task "hello" --ttl-s 60
//   python3 serve.py --check /path/to/spool/demo.json   # -> check passed: agent file is valid
//
// Same CLI as emit.py. Writes <spool>/<id>.json atomically: writeFileSync to a
// temp name in the same dir (not ending in .json), fsyncSync, renameSync.
// Node >= 18, no dependencies.

import { mkdirSync, writeFileSync, openSync, fsyncSync, closeSync, renameSync, unlinkSync } from "node:fs";
import { randomBytes } from "node:crypto";

function die(message) {
  process.stderr.write(`emit: ${message}\n`);
  process.exit(2);
}

function parseArgs(argv) {
  const args = {
    path: null,
    id: null,
    name: null,
    emoji: null,
    role: null,
    model: null,
    task: null,
    blocked: null,
    needs_decision: null,
    done: null,
    ctx_pct: null,
    log: [],
    ttl_s: null,
  };
  const take = (flag, i) => {
    if (i + 1 >= argv.length) die(`missing value for ${flag}`);
    return argv[i + 1];
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    switch (arg) {
      case "--id": args.id = take(arg, i); i++; break;
      case "--name": args.name = take(arg, i); i++; break;
      case "--emoji": args.emoji = take(arg, i); i++; break;
      case "--role": args.role = take(arg, i); i++; break;
      case "--model": args.model = take(arg, i); i++; break;
      case "--task": args.task = take(arg, i); i++; break;
      case "--blocked": args.blocked = take(arg, i); i++; break;
      case "--needs-decision": args.needs_decision = take(arg, i); i++; break;
      case "--done": args.done = take(arg, i); i++; break;
      case "--ctx-pct": args.ctx_pct = take(arg, i); i++; break;
      case "--log": args.log.push(take(arg, i)); i++; break;
      case "--ttl-s": args.ttl_s = take(arg, i); i++; break;
      case "--": break;
      default:
        if (arg.startsWith("-")) die(`unknown option: ${arg}`);
        if (args.path !== null) die(`unexpected argument: ${arg}`);
        args.path = arg;
    }
  }
  return args;
}

function isInt(value) {
  return typeof value === "string" && /^\d+$/.test(value);
}

function parseDone(value) {
  const eq = value.indexOf("=");
  if (eq !== -1) return { label: value.slice(0, eq), url: value.slice(eq + 1) };
  return value;
}

function writeAgentFile(pathDir, record) {
  mkdirSync(pathDir, { recursive: true });
  const finalPath = `${pathDir}/${record.id}.json`;
  const payload = JSON.stringify(record);
  const tmpName = `.${record.id}.json.${randomBytes(6).toString("hex")}`;
  const tmpPath = `${pathDir}/${tmpName}`;
  let fd;
  try {
    fd = openSync(tmpPath, "w");
    writeFileSync(fd, payload, "utf8");
    fsyncSync(fd);
    closeSync(fd);
    renameSync(tmpPath, finalPath);
  } catch (err) {
    try { unlinkSync(tmpPath); } catch { /* already gone */ }
    throw err;
  }
  return finalPath;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.path === null) die("missing spool directory argument");
  if (args.id === null) die("--id is required");

  if (args.ctx_pct !== null) {
    if (!isInt(args.ctx_pct) || Number(args.ctx_pct) > 100) {
      die(`--ctx-pct must be an integer 0..100, got ${args.ctx_pct}`);
    }
  }
  if (args.ttl_s !== null) {
    if (!isInt(args.ttl_s) || Number(args.ttl_s) < 1) {
      die(`--ttl-s must be a positive integer, got ${args.ttl_s}`);
    }
  }

  const record = { office: 1, id: args.id };
  for (const field of ["name", "emoji", "role", "model", "task", "blocked", "needs_decision"]) {
    if (args[field] !== null) record[field] = args[field];
  }
  if (args.done !== null) record.done = parseDone(args.done);
  if (args.ctx_pct !== null) record.ctx_pct = Number(args.ctx_pct);
  if (args.log.length > 0) record.log = args.log;
  if (args.ttl_s !== null) record.ttl_s = Number(args.ttl_s);

  writeAgentFile(args.path, record);
}

main();

File: examples/agent-file/emit.py
"""Quickstart: emit one valid agent spool file (schema v1).

    python3 examples/agent-file/emit.py /path/to/spool --id demo --task "hello" --ttl-s 60
    python3 serve.py --check /path/to/spool/demo.json   # -> check passed: agent file is valid

Writes <spool>/<id>.json atomically (temp file in the same dir + fsync + os.replace); stdlib only.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import tempfile
from pathlib import Path

_STRING_FIELDS = (
    ("name", "name"),
    ("emoji", "emoji"),
    ("role", "role"),
    ("model", "model"),
    ("task", "task"),
    ("blocked", "blocked"),
    ("needs_decision", "needs_decision"),
)


def write_agent_file(path_dir, record: dict) -> Path:
    """Atomically write *record* to ``<path_dir>/<record['id']>.json``.

    Serializes the complete object to a temporary file in the same directory
    whose name does not end in ``.json``, flushes and ``os.fsync``s it, then
    ``os.replace``s it onto the final path. Returns the final ``Path``.
    """
    directory = Path(path_dir)
    directory.mkdir(parents=True, exist_ok=True)
    agent_id = record["id"]
    final = directory / f"{agent_id}.json"
    payload = json.dumps(record, ensure_ascii=False)
    fd, tmp_name = tempfile.mkstemp(
        prefix=f".{agent_id}.json.", suffix=".tmp", dir=directory
    )
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(payload)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(tmp_name, final)
    except BaseException:
        try:
            os.unlink(tmp_name)
        except OSError:
            pass
        raise
    return final


def _parse_done(value: str):
    """``LABEL`` -> string; ``LABEL=URL`` -> ``{"label": ..., "url": ...}``."""
    if "=" in value:
        label, url = value.split("=", 1)
        return {"label": label, "url": url}
    return value


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(
        description="Emit one valid agent spool file (schema v1)."
    )
    parser.add_argument("path", help="spool directory (created if missing)")
    parser.add_argument("--id", required=True, dest="id", help="agent id (filename stem)")
    parser.add_argument("--name", help="display name")
    parser.add_argument("--emoji", help="badge emoji")
    parser.add_argument("--role", help="role (room_for() input)")
    parser.add_argument("--model", help="collar-chip model label")
    parser.add_argument("--task", help="one-line current-work claim")
    parser.add_argument("--blocked", help="blocker text")
    parser.add_argument("--needs-decision", dest="needs_decision", help="question text")
    parser.add_argument("--done", help="LABEL or LABEL=URL")
    parser.add_argument("--ctx-pct", dest="ctx_pct", type=int, help="context used, 0..100")
    parser.add_argument("--log", action="append", help="log line (repeatable)")
    parser.add_argument("--ttl-s", dest="ttl_s", type=int, help="positive integer seconds")
    args = parser.parse_args(argv)

    if args.ctx_pct is not None and not 0 <= args.ctx_pct <= 100:
        print(f"emit: --ctx-pct must be an integer 0..100, got {args.ctx_pct}", file=sys.stderr)
        return 2
    if args.ttl_s is not None and args.ttl_s <= 0:
        print(f"emit: --ttl-s must be a positive integer, got {args.ttl_s}", file=sys.stderr)
        return 2

    record = {"office": 1, "id": args.id}
    for flag, field in _STRING_FIELDS:
        value = getattr(args, flag)
        if value is not None:
            record[field] = value
    if args.done is not None:
        record["done"] = _parse_done(args.done)
    if args.ctx_pct is not None:
        record["ctx_pct"] = args.ctx_pct
    if args.log:
        record["log"] = list(args.log)
    if args.ttl_s is not None:
        record["ttl_s"] = args.ttl_s

    write_agent_file(args.path, record)
    return 0


if __name__ == "__main__":
    sys.exit(main())
